presto-orc/src/main/java/com/facebook/presto/orc/reader/StructStreamReader.java

```diff
@@ -96,18 +97,28 @@ public Block readBlock(Type type)
         Block[] blocks = new Block[typeParameters.size()];
         if (presentStream == null) {
             for (int i = 0; i < typeParameters.size(); i++) {
-                StreamReader structField = structFields[i];
-                structField.prepareNextRead(nextBatchSize);
-                blocks[i] = structField.readBlock(typeParameters.get(i));
+                if (i < structFields.length) {
+                    StreamReader structField = structFields[i];
+                    structField.prepareNextRead(nextBatchSize);
+                    blocks[i] = structField.readBlock(typeParameters.get(i));
+                }
+                else {
+                    blocks[i] = getNullBlock(typeParameters.get(i));
+                }
             }
         }
         else {
             int nullValues = presentStream.getUnsetBits(nextBatchSize, nullVector);
             if (nullValues != nextBatchSize) {
                 for (int i = 0; i < typeParameters.size(); i++) {
-                    StreamReader structField = structFields[i];
-                    structField.prepareNextRead(nextBatchSize - nullValues);
-                    blocks[i] = structField.readBlock(typeParameters.get(i));
+                    if (i < structFields.length) {
+                        StreamReader structField = structFields[i];
+                        structField.prepareNextRead(nextBatchSize - nullValues);
+                        blocks[i] = structField.readBlock(typeParameters.get(i));
+                    }
+                    else {
+                        blocks[i] = getNullBlock(typeParameters.get(i));
```

Comment:
This will create a block with one null entry per struct, but some of the structs are null.  Instead it should create a block with `nextBatchSize - nullValues` entries.   We can catch this kind of mistake with this after the `if/else`:
```java
        verify(Arrays.stream(blocks)
                .mapToInt(Block::getPositionCount)
                .distinct()
                .count() == 1);
```